src/webgpu/shader/validation/uniformity/uniformity.spec.ts

```diff
@@ -1828,3 +1828,258 @@ fn main() {
+    function: `fn foo() -> vec4f {
+      return textureSample(t,s,vec2f(0,0));
+    }`,
+    call: `if nonuniform_cond {
+      let call = foo();
+    }`,
+    cond: `uniform_cond`,
+    uniform: false,
+  },
+  required_uniform_function_call_in_nonuniform_cf2: {
+    function: `@diagnostic(warning, derivative_uniformity)
+    fn foo() -> vec4f {
+      return textureSample(t,s,vec2f(0,0));
+    }`,
+    call: `if nonuniform_cond {
+      let call = foo();
+      let sample = textureSample(t,s,vec2f(0,0));
+    }`,
+    cond: `uniform_cond`,
+    uniform: false,
+  },
+  required_uniform_function_call_depends_on_uniform_param: {
+    function: `fn foo(x : bool) -> vec4f {
+      if x {
+        return textureSample(t,s,vec2f(0,0));
+      }
+      return vec4f(0);
+    }`,
+    call: `let call = foo(uniform_cond);`,
+    cond: `uniform_cond`,
+    uniform: true,
+  },
+  required_uniform_function_call_depends_on_nonuniform_param: {
+    function: `fn foo(x : bool) -> vec4f {
+      if x {
+        return textureSample(t,s,vec2f(0,0));
+      }
+      return vec4f(0);
+    }`,
+    call: `let call = foo(nonuniform_cond);`,
+    cond: `uniform_cond`,
+    uniform: false,
+  },
+  dpdx_nonuniform_result: {
+    function: ``,
+    call: `let call = dpdx(1);`,
+    cond: `call > 0`,
+    uniform: false,
+  },
+  dpdy_nonuniform_result: {
+    function: ``,
+    call: `let call = dpdy(1);`,
+    cond: `call > 0`,
+    uniform: false,
+  },
+  dpdxCoarse_nonuniform_result: {
+    function: ``,
+    call: `let call = dpdxCoarse(1);`,
+    cond: `call > 0`,
+    uniform: false,
+  },
+  dpdyCoarse_nonuniform_result: {
+    function: ``,
+    call: `let call = dpdyCoarse(1);`,
+    cond: `call > 0`,
+    uniform: false,
+  },
+  dpdxFine_nonuniform_result: {
+    function: ``,
+    call: `let call = dpdxFine(1);`,
+    cond: `call > 0`,
+    uniform: false,
+  },
+  dpdyFine_nonuniform_result: {
+    function: ``,
+    call: `let call = dpdyFine(1);`,
+    cond: `call > 0`,
+    uniform: false,
+  },
+  fwidth_nonuniform_result: {
+    function: ``,
+    call: `let call = fwidth(1);`,
+    cond: `call > 0`,
+    uniform: false,
+  },
+  fwidthCoarse_nonuniform_result: {
+    function: ``,
+    call: `let call = fwidthCoarse(1);`,
+    cond: `call > 0`,
+    uniform: false,
+  },
+  fwidthFine_nonuniform_result: {
+    function: ``,
+    call: `let call = fwidthFine(1);`,
+    cond: `call > 0`,
+    uniform: false,
+  },
+  textureSample_nonuniform_result: {
+    function: ``,
+    call: `let call = textureSample(t,s,vec2f(0,0));`,
+    cond: `call.x > 0`,
+    uniform: false,
+  },
+  textureSampleBias_nonuniform_result: {
+    function: ``,
+    call: `let call = textureSampleBias(t,s,vec2f(0,0), 0);`,
+    cond: `call.x > 0`,
+    uniform: false,
+  },
+  textureSampleCompare_nonuniform_result: {
+    function: ``,
+    call: `let call = textureSampleCompare(td,sd,vec2f(0,0), 0);`,
+    cond: `call > 0`,
+    uniform: false,
+  },
```

Comment:
Suggest also testing that other texture builtins like `textureSampleLevel` and `textureLoad` return a uniform value iff the inputs are all uniform.

And maybe one or two other builtin functions too (e.g. math builtins, type constructors/conversions, bitcast)? You've got good coverage for `ParameterReturnContentsRequiredToBeUniform` for user functions, but not for builtins.

Reply:
Added in the latest commit.